.github/workflows/test.yml

```diff
@@ -1,32 +1,54 @@
+
 name: build
+
 on:
+  pull_request:
+    types: [ opened, synchronize ]
   push:
-    branches: '*'
+    branches: [ master ]
+
 jobs:
   run:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
         os: [ubuntu-latest]
+        python-version: [ 3.9 ]
     env:
       OS: ${{ matrix.os }}
-      PYTHON: '3.9'
+      PYTHON: ${{ matrix.python-version }}
+
     steps:
-    - uses: actions/checkout@master
-    - name: Setup Python
+
+    - name: Checkout code
+      uses: actions/checkout@master
+
+    - name: Set up Python ${{ matrix.python-version }}
       uses: actions/setup-python@master
       with:
-        python-version: 3.9
+        python-version: ${{ matrix.python-version }}
+
+    - name: Install Dependencies
+      working-directory: ./python
+      run: |
+        python -m pip install --upgrade pip
+        pip install -U -r requirements.txt
+
     - name: Set Spark env
       run: |
         export SPARK_LOCAL_IP=127.0.0.1
         export SPARK_SUBMIT_OPTS="--illegal-access=permit -Dio.netty.tryReflectionSetAccessible=true"
+
+    - name: Unit Tests
+      working-directory: ./python
+      run: python -m unittest discover -p "*_tests.py"
```

Comment:
@tnixon 
Question: Is there any way that we can only run the newly added test codes only?

Reply:
Are we not running the tests twice if we include this block?

Reply:
Hmmm - maybe I'm not understanding fully how codecov works. I want any PR to be blocked if there are failing unit tests, so not sure if codecov will achieve this alone, which is why I added this additional run.

I understand that codecov runs the tests in order to evaluate coverage, but will it fail if one of the underlying tests fail, or does it just produce the report?

Reply:
@souvik-databricks - the goal here is to check whether changes have broken any existing code, not just to run newly developed test code. We want to prevent regressions (breaking things that were working previously).